### PR TITLE
Port Vision Zero import process to Prefect

### PR DESCRIPTION
**Pull request by frankhereford**
## Associated Issues

This PR intends to close https://github.com/cityofaustin/atd-data-tech/issues/8861.

## TLDR

This PR is the result of a mini-spike to take another look at how we might want to structure our prefect flows, especially with regard to side-stepping the storage requirements and use of docker. 🐳 

## Testing notes

As configured, the flow will grab the development KV store from prefect. This configures the VZ import process to download any archives from our SFTP endpoint and load them into a postgres/graphql-engine pair I have running on `atd-data03`. I opted not to use staging to give myself total freedom to empty out tables in the VZ DB to easily isolate changes. Please feel free to refer to how I did that and use / stop / examine my docker-compose based VZ dev stack.

**NB:** This configuration means you can run this without worry that you'll mess up the VZDB, but you should be cognizant that this script will delete archives off the SFTP endpoint, and those archives will **not** be available to be imported into the production VZDB. Please back them up and return them to `/home/txdot/` or make arrangements for the data to be backfilled. I would be happy to sync with anyone who needs this performed or would like to learn more about how the VZ import process works.

As for more detailed testing instructions, I am actually inclined to not try to define them. This PR is more about showing one way (out of multitudes) that a flow can be written and executed, and I want to invite you to spend review time not so much hunting bugs but getting to know the code and seeing how it functions. Of course, any bug reports are welcome as well. 🐛 

If this structure does make it out of the gate and it looks like we'll use it (at least for VZ's import Flow), I'll write up documentation on how to stand it up in the README if what is there doesn't suffice. As it stands, that README.md is a great starting place for testing too.

## TODO
* [x] Add archive task for zip files received
* [x] Remove archives from sftp endpoint `ssh rm` on completion
* [x] Document all the not-checked-in-files
* [x] Write a README.md
* [x] Remove all debug comments
* [ ] ~Artifacts~
* [x] Add restart always to docker-compose file
* [x] Convert all / some env variables to cloud UI held secrets

## Stretch
* [x] Parallelize non-crash object import steps

### CODE REVIEW

File: flows/vision-zero/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+The same docker image is used to develop the flows. Because the image is built to include the libraries that the project's flows may require, there is a intrinisic guarentee that if the flow's package needs are met during development, they will be similarly met during agent execution. This docker contained development execution mechanism also sidesteps the need to define and manage a Flow storage solution allowing the registration process to convey the actual code to be run directly to the agent via the default local storage.
+
+```
+# you can connect to the agent / flow development container 
+$ docker exec -it prefect-agent-vz bash
+```
+
+## Environment Configuration
+
+Environment variables are defined via an environemnt file that is passed to docker when the container is started and which is specificed in the `docker-compose.yml` file. Ultimately, these need to be refactored into a single JSON blob and defined for each deployment phase in the application life cycle (dev, staging, production). This work should be done in conjunction with deploying this code once the group is in agreement over the methods used in this flow/agent setup.
```
Comment (chiaberry) [inline flows/vision-zero/README.md:35]: typo: environment

Comment (frankhereford) [reply flows/vision-zero/README.md]: Thanks @chiaberry, just pushed a fix for that. 🙏

File: flows/vision-zero/cris_import/cris_import.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    rsync = sysrsync.run(
+        verbose=True,
+        options=["-a"],
+        source_ssh=SFTP_ENDPOINT,
+        source="/home/txdot/*zip",
+        sync_source_contents=False,
+        destination=zip_tmpdir,
+    )
+    logger.info("Rsync return code: " + str(rsync.returncode))
+    if rsync.returncode != 0:
```
Comment (mddilley) [inline flows/vision-zero/cris_import/cris_import.py:108]: This stuck out as another place for a comment. A quick Google search got me to the answer that 0 = success, and I think that is could be argued that it can be inferred from the code as well. Just an observation.

**Review by mddilley — COMMENTED** [review]
I made first pass through the code and have some questions. I realize that some of this code already existed so I apologize if I dug a little too much into readability. I'm still going to spend some more time on this as I learn more about Prefect and this Docker setup.

File: flows/vision-zero/cris_import/make_kv_store.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/python3
+
+import os
+import json
+from prefect.backend import set_key_value
+
+RAW_AIRFLOW_CONFIG_JSON = os.getenv("RAW_AIRFLOW_CONFIG")
```
[the remark this answers is not in the document]
Comment (mddilley) [reply flows/vision-zero/cris_import/make_kv_store.py:7]: Thanks for addressing this in the Prefect Huddle today. 🙏 I'll log in to Prefect Cloud asap and check out the env vars.

File: flows/vision-zero/cris_import/cris_import.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+def download_extract_archives():
+    """
+    Connect to the SFTP endpoint which receives archives from CRIS and
+    download them into a temporary directory.
+
+    Returns path of temporary directory as a string
+    """
+
+    logger = prefect.context.get("logger")
+    logger.info(sys._getframe().f_code.co_name + "()")
```
[the remark this answers is not in the document]
Comment (frankhereford) [reply flows/vision-zero/cris_import/cris_import.py:97]: Thanks for asking this Mike. I put this code at the top of each of my tasks during development, and specifically when I was working with the prefect's two ways of defining the dependencies between tasks. They have a functional method, where you pass and return things, and prefect is able to determine inter-task dependencies, and they have an explicit method, where explicitly say what depends on what. 

In the course of getting all that straight in my head, it became very helpful for me to have a print statement at the top of each task that would say "Hey, we're inside this task now." I even had them colored green with ANSI codes for a while, when it was all in my terminal. 

All this to say, these "hey, we're in here" debug statements can go, as the prefect ecosystem gives us a much more usable, visual, graph-as-in-graph-theory view via the cloud UI. This is vestigial from my early development. 

I will strike them in a forthcoming commit.

File: flows/vision-zero/cris_import/cris_import.py [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+    charges_import_tmpdirs = run_docker_image.map(
+        extracted_data=uploaded_archives_csvs,
+        vz_etl_image=unmapped(image),
+        command=unmapped(["/app/process_hasura_import.py", "charges"]),
+    )
+    charges_import_tmpdirs.set_upstream(crash_import_tmpdirs)
+
+    # remove archives from SFTP endpoint
+    removal_token = remove_archives_from_sftp_endpoint(zip_location)
+    removal_token.set_upstream(crash_import_tmpdirs)
```
[the remark this answers is not in the document]
Comment (frankhereford) [reply flows/vision-zero/cris_import/cris_import.py:358]: You got it. This is the ["explicit method"](https://github.com/cityofaustin/atd-prefect/pull/9#discussion_r891757784) of defining task interdependencies that I am referring to in this comment.

File: flows/vision-zero/cris_import/cris_import.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mddilley) [reply flows/vision-zero/cris_import/cris_import.py:97]: Cool, thanks for the explanation. That makes total sense to me. 🙏

File: flows/vision-zero/cris_import/cris_import.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+
+    # get a location on disk which contains the zips from the sftp endpoint
+    zip_location = download_extract_archives()
+
+    # iterate over the zips in that location and unarchive them into
+    # a list of temporary directories containtain the files of each
+    extracted_archives = unzip_archives(zip_location)
+
+    # push up the archives to s3 for archival
+    uploaded_archives_csvs = upload_csv_files_to_s3.map(extracted_archives)
```
[the remark this answers is not in the document]
Comment (frankhereford) [reply flows/vision-zero/cris_import/cris_import.py:318]: Totally - `map`ping is key to being able to parallelize multiple paths through the flow of sets of similar data / tasks. The forthcoming version of prefect re-works this to be even more pythony, and buries this functionality inside "normal" python syntax I believe.

File: flows/vision-zero/cris_import/cris_import.py (hunk 2 above)
Comment (frankhereford) [reply flows/vision-zero/cris_import/cris_import.py:108]: I can totally get behind that. I appreciate the suggestion; it didn't occur to me because `0` is default return code for programs written in C or maybe more generally for system software on a POSIX host. Noting that in the code is totally helpful, and I will add that. Thank you!

**Comment by frankhereford** [comment]
@mddilley wrote:
> I made first pass through the code and have some questions. I realize that some of this code already existed so I apologize if I dug a little too much into readability. I'm still going to spend some more time on this as I learn more about Prefect and this Docker setup.

Mike - thank you for reviewing the code and for your comments! I appreciate your perspective and questions a ton, and I have put up some replies. Let me know if there's anything else that comes to mind, or if my replies are not as helpful as they could be. 

And specifically to your comments about readability -- they are a million times welcome. Thank you!!

File: flows/vision-zero/cris_import/cris_import.py [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+
+    # remove archives from SFTP endpoint
+    removal_token = remove_archives_from_sftp_endpoint(zip_location)
+    removal_token.set_upstream(crash_import_tmpdirs)
+    removal_token.set_upstream(unit_import_tmpdirs)
+    removal_token.set_upstream(person_import_tmpdirs)
+    removal_token.set_upstream(primaryperson_import_tmpdirs)
+    removal_token.set_upstream(charges_import_tmpdirs)
+
+    cleanup = cleanup_temporary_directories(
```
[the remark this answers is not in the document]
Comment (mddilley) [reply flows/vision-zero/cris_import/cris_import.py:364]: Also wanted to note that I was happy to hear that some the processes that would be hard to know were even running were pulled into this flow. Very cool to make them more visible as part of this port.

[the remark this answers is not in the document]
Comment (frankhereford) [reply flows/vision-zero/cris_import/cris_import.py:356]: @mddilley wrote: 
> Was this the part that you were referring that was running elsewhere? @frankhereford

Yep, you are correct. I'm referring to the cronjob that runs on the SFTP endpoint that daily unzips any zip files it's received and uploads the resulting CSV files to airflow. This was not part of the airflow ETL, but was a behavior that the airflow ETL depended on. 

It's because I short-circuit the uploading of CSV files by the SFTP endpoint, that this flow takes on the responsibility of uploading the CSV files to S3 for archival -- so we do get that permanent copy which we are used to.

**Review by mddilley — APPROVED** [review]
I am all aboard the dockerized-agent train! 🚂 💯 

I'll get some notes together so I can talk about my experience using this for dev so far.

**Review by patrickm02L — APPROVED** [review]
Very excited for this! 🤘

File: flows/vision-zero/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ENV TZ="America/Chicago"
+RUN apt update
+RUN apt-get update && \
+    apt-get install -yq tzdata && \
+    ln -fs /usr/share/zoneinfo/America/Chicago /etc/localtime && \
+    dpkg-reconfigure -f noninteractive tzdata
+RUN apt upgrade -y
+RUN apt install -y vim magic-wormhole aptitude python3-pip docker.io rsync p7zip-full black
+RUN pip install prefect
+RUN pip install sysrsync GitPython boto3
```
Comment (johnclary) [inline flows/vision-zero/Dockerfile:11]: @frankhereford can these `pip install` commands be combined into a single `requirements.txt` file (with pinned version numbers), and installed here with ` RUN pip install -r requirements.txt`?

**Review by johnclary — APPROVED** [review]

File: flows/vision-zero/Dockerfile (hunk 8 above)
Comment (frankhereford) [reply flows/vision-zero/Dockerfile:11]: @johnclary wrote: 
> @frankhereford can these pip install commands be combined into a single requirements.txt file (with pinned version numbers), and installed here with RUN pip install -r requirements.txt?

Broadly yes, I think it's more pythonic to utilize pip's ability to freeze out an environment and install from the requirements.txt which is generated. However, it's a little bit tricky here because the Dockerfile is defining the environment for the agent which can run multiple flows, and I envision each flow having a requirements.txt. I will investigate a way to have the Dockerfile iterate over each flow's requirements.txt and install from there. 

I'm not sure how much this matters, but one thing the docker file's `pip` invocations are doing is installing these as system-wide available python libraries within the dockerized linux host and not into a encapsulated `venv`. 

I am going to spin off an issue to address this, and I hope maybe you can I can sync on what I come up with or what I plan to come up with, depending on how it goes.

Comment (johnclary) [reply flows/vision-zero/Dockerfile:11]: OK, I think I misunderstood. I thought the per-agent Dockerfile was meant to provide an environment for all flows that this agent would run. As you add more flows to this agent's scope (e.g. the narrative extractor), do those requirements get stuffed somewhere else?

FWIW if it makes life easier I think it's fine to let one requirements.txt grow to support all of an agent's tasks. If an agent is generally managing roughly a repo's worth of ETLs, it would hopefully be rare for us to run into major dependency conflicts.
